src/Raven.Server/Storage/Schema/Updates/Documents/50000/From42017.cs

```diff
@@ -1,16 +1,493 @@
-﻿namespace Raven.Server.Storage.Schema.Updates.Documents
+﻿using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using Raven.Client.Documents.Changes;
+using Raven.Server.Documents;
+using Raven.Server.Documents.Replication.ReplicationItems;
+using Raven.Server.ServerWide.Context;
+using Raven.Server.Utils;
+using Sparrow;
+using Sparrow.Binary;
+using Sparrow.Json;
+using Sparrow.Json.Parsing;
+using Sparrow.Server;
+using Sparrow.Server.Utils;
+using Voron;
+using Voron.Data.Tables;
+using static Raven.Server.Documents.DocumentsStorage;
+using static Raven.Server.Documents.CountersStorage;
+
+
+namespace Raven.Server.Storage.Schema.Updates.Documents
 {
-    public unsafe class From42017 : ISchemaUpdate
+    public class From42017 : ISchemaUpdate
     {
         public int From => 42_017;
 
         public int To => 50_000;
 
         public SchemaUpgrader.StorageType StorageType => SchemaUpgrader.StorageType.Documents;
 
+        internal static int NumberOfCounterGroupsToMigrateInSingleTransaction = 10_000;
+
+        private string _dbId;
+
+        private class CounterBatchUpdate
+        {
+            public readonly List<CounterReplicationItem> Counters = new List<CounterReplicationItem>();
+
+            public void Clear()
+            {
+                foreach (var counterGroup in Counters)
+                {
+                    counterGroup.Dispose();
+                }
+
+                Counters.Clear();
+            }
+        }
+
         public bool Update(UpdateStep step)
         {
+            var readTable = new Table(CountersSchema, step.WriteTx);
+            var countersTree = readTable.GetTree(CountersSchema.Key);
+            if (countersTree == null)
+                return true;
+            
+            step.DocumentsStorage.CountersStorage = new CountersStorage(step.DocumentsStorage.DocumentDatabase, step.WriteTx);
+            
+            // this schema update uses DocumentsStorage.GenerateNextEtag() so we need to read and set LastEtag in storage
+            _dbId = From41016.ReadDbId(step);
+            step.DocumentsStorage.InitializeLastEtag(step.ReadTx);
+
+            var toDeleteByCollection = new Dictionary<string, List<long>>();
+            string currentDocId = null;
+            string currentPrefix = null;
+            var batch = new CounterBatchUpdate();
+            var done = false;
+
+            // 4.2 counters processing
+
+            while (done == false)
+            {
+                readTable = new Table(CountersSchema, step.ReadTx);
+
+                var processedInCurrentTx = 0;
+
+                using (step.DocumentsStorage.ContextPool.AllocateOperationContext(out DocumentsOperationContext context))
+                {
+                    var commit = false;
+
+                    foreach (var counterGroup in GetCounters(readTable, context, currentPrefix))
+                    {
```

Comment:
We recreate here all counters, regardless of their casing?

Reply:
Yes, we delete all `CounterGroup` entries and recreate them all from scratch
We can't simply lower-case the counter names (and keys) in each CounterGroup entry, 
because lower-casing a counter name changes it's lexical value and can affect the total ordering between CounterGroup entries.

e.g. say we have 2 CounterGroup entries :
- [A, B, X, Y] (key = "users/1|")
- [c, d, e, f] (key = "users/1|c")

if we were to lower the casing in each group we will get - 
- [a, b, x, y] (key = "users/1|")
- [c, d, e, f] (key = "users/1|c")

And now we are in a bad state - CounterGroup entry "users/1|" comes before "users/1|c" entry,
but "users/1|" entry contains counter names with lexical value that is greater than the first counter in the next entry ("x" and "y" are greater than "c")
so now, for example if we want to get value of counter "x", we will search for it in entry "users/1|c" (because that's where we think it's supposed to be), won't find it there and return null.. Similar problem with increment